Cli parsing

lpenzey opened the pull request:
This feature incorporates JCommander to parse command line arguments to ensure a port is given in the correct manner, meaning that it's an integer between 1024 and 49151 in order to stay away from system/dynamic ports. Entering no arguments or `--help` will display usage.

lpenzey committed:
Make logger public

lpenzey committed:
Add jcommander dependency to gradle build

lpenzey committed:
Incorporate command line parsing/validation

lpenzey committed:
Define CL args and validation

Args defines --port as a required argument when starting the program. It
is set to accept an integer between 1024 and 49151 inclusive so as not
to interfere with system ports (under 1024) and dynamic ports (over
49151). If no argument is provided it will display usage to the user.
If a non integer is given it will exit and display a prompt to the user.
Users can also pass the --help argument to get the usage.

lpenzey committed:
Update README

lpenzey commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Server - CLI Flag Parsing - 1](https://trello.com/c/OauR63ua/71-server-cli-flag-parsing-1)

ssunday commented on src/main/java/server/CLI/PortValidator.java:
```diff
@@ -0,0 +1,24 @@
+        }
+
+        if (n < 1024 || n > 49151)
+            throw new ParameterException("Parameter " + name + " should be a number between 1024 and 49151 (found " + value + ")");
```
...is there any way to validate that that port is not being locked or used by something else? That's like extra credit or something, just a thought

lpenzey replied:
That would be pretty cool, I'll look into it

lpenzey committed:
Delete outdated handlers

lpenzey committed:
Convert Handler to functional interface

Handler is a functional interface in order to take advantage of lambdas
as callbacks. A Handler is created to give custom functionality to
responding to routes defined by the application.

When a handler is invoked it sends the request (and body if applicable"
to HandlerHelper, which creates a new ResponseBuilder and sends the
necessary information to build the correct response.

lpenzey committed:
Separate application from server

App contains defined routes, which are built in the Routes class. The
routes are stored as a TreeMap in order to easily accommodate an unknown
number of routes. The routes in the TreeMap correlate with a HashMap
that contains the allowed methods and the corresponding handler.

Once the routes are defined, App is injected into the Server to allow
for the server to check incoming requests against known routes housed in
the App.

lpenzey committed:
Convert router to hold routes in a TreeMap

lpenzey committed:
Remove setters from StatusLine

lpenzey committed:
Add PathFinder to create response for request path

PathFinder checks the incoming request for a matching path in the routes
defined by the app. If there's a match, the response creation will be
delegated to the handler that matches the found path/route.

If a match is not found then the response creation gets handed to
ResponseBuilder to return a 404 Not Found Response. Similarly, if a
matching path is found but is being accessed by an incorrect method,
ResponseBuilder will return a Method Not Allowed response.

lpenzey committed:
Add intention revealing variables

lpenzey committed:
Replace setters with constructor parameters

lpenzey committed:
Inject app and routes to server

lpenzey committed:
Add ResponseBuilder to centralize response creation

lpenzey committed:
Add deployment settings to travis

lpenzey committed:
Fix syntax in Procfile

lpenzey committed:
Merge pull request #6 from lpenzey/continuous_deployment

Add deployment settings to travis

lpenzey committed:
Merge pull request #5 from lpenzey/routing_refactor

Routing refactor
